### PR TITLE
Improve bounds checking error

### PR DESCRIPTION
**Pull request by mppf**
This PR does 2 things related to array bounds checking:
 1. It improves the error message for out-of-bounds indexing or slicing
    to include details of the array bounds and to try to present this
    information in a manner that is understandable.
 2. It migrates the bounds check on access for arrays over rectangular
    domains to the relevant functions in ChapelArray (e.g.
    `_array.this()`, `_array.localAccess()`) instead of having these
    implemented per-domain in `dsiAccess`. The main reason to do this is
    that it's easier to know that all DSI implementations rectangular
    arrays have bounds checking (vs. before we might miss one in the
    various dsiAccess/LocalAccess functions). I started to try it because
    I noticed that slices were bounds-checked in ChapelArray but element
    access was not.

This PR changed the behavior of two futures. I created an issue to
discuss each future and described how the PR changed the future in those
issues. Please see #14264 and #14264.

Reviewed by @benharsh - thanks!
Thanks also to @lydia-duncan for help with the specific error message.

- [x] full local testing
- [x] passed gasnet testing

### CODE REVIEW

**Comment by mppf** [comment]
arrays/bradc/localSlices/localsliceOOB.chpl
arrays/bradc/localSlices/localsliceOOB2.chpl
arrays/bradc/localSlices/nonlocalslice.chpl
arrays/bradc/slices/outOfBoundsSlice.chpl
arrays/bradc/slices/sliceArray.chpl
arrays/bradc/slices/sliceStridedArray.chpl
arrays/deitz/part6/test_bounds1.chpl
arrays/deitz/part6/test_bounds2.chpl
arrays/waynew/init2.chpl
distributions/bharshbarg/stencil/localReads.chpl
distributions/bharshbarg/stencil/periodic.chpl
distributions/bharshbarg/stencil/reshape.chpl
distributions/bharshbarg/stencil/slice.chpl
distributions/bharshbarg/stencil/small.chpl
distributions/bradc/blkCyc/blkCycSliceOOB.chpl
distributions/bradc/idxToLocale.chpl
execflags/bradc/gdbddash/declint2.chpl
library/packages/UnitTest/Launcher_Test.chpl
multilocale/bradc/referToLocaleOne.chpl
optimizations/forall-unordered-opt/compilation-tests/blocking-analysis2.chpl
optimizations/forall-unordered-opt/compilation-tests/blocking-analysis3.chpl
optimizations/forall-unordered-opt/compilation-tests/blocking-analysis4.chpl
optimizations/forall-unordered-opt/compilation-tests/opt-expected.chpl
release/examples/benchmarks/miniMD/miniMD.chpl
studies/bale/histogram/histo-atomics-forall-opt.chpl
users/npadmana/npb-mg/mg.chpl

**Comment by mppf** [comment]
Seeing these failures in gasnet testing:

[Error matching program output for arrays/slices/commCounts/sliceBlockWithRanges]
[Error matching program output for arrays/slices/commCounts/sliceLocDist]
[Error matching program output for llvm/tbaa/tbaa_scalar]
[Error matching program output for modules/sungeun/init/printInitCommCounts]
[Error matching program output for performance/ferguson/array-assign-block-get]
[Error matching program output for performance/hilde/coforallOnByRef]
[Error matching program output for performance/hilde/userSync]
[Error matching program output for studies/hpcc/CommDiags/stream-ep_commDiags]

File: modules/internal/ChapelArray.chpl [hunk 1]
```diff
@@ -2361,37 +2361,179 @@ module ChapelArray {
...
+                 "note: ", dimstr);
+          }
+        }
+      }
+    }
+
+    pragma "insert line file info"
+    pragma "always propagate line file info"
+    pragma "no doc"
+    proc checkSlice(ranges...rank, value) where chpl__isTupleOfRanges(ranges) {
```
Comment (benharsh) [inline modules/internal/ChapelArray.chpl]: Seems like we could re-use this code for the domain case. Could you call ``d.dsiDims()`` or create the tuple yourself, then pass it to this method?

Comment (mppf) [reply modules/internal/ChapelArray.chpl]: I agree and actually I had a version that worked that way originally. I had tried using `d.getIndices()` but that did not function correctly for sparse domains. It didn't occur to me to use `d.dsiDims()` though - which seems to work - so thanks for the tip!

**Pull request merged**